### PR TITLE
Add symfony mailer integration

### PR DESCRIPTION
**Pull request by bramtervoort**
| Q | A
| --- | ---
| Bug fix? | no
| New feature? | yes
| BC breaks? | no
| Deprecations? | no
| Fixed tickets | fixes #issuenum
| Related issues/PRs | #issuenum
| License | MIT

#### What's in this PR?

Created a mailer helper class to send emails using the new sf mailer
component. Created a configuration option to switch to varying mail
helpers.

#### Why?

It would be nice to be able to use the new Symfony Mailer component for sending the emails. 

#### Example Usage

In the configuration just state: 
```yml
sulu_form:
    mail:
        helper: "mailer"
```

Note that the user can now create customer helpers to and wire them in using a tag and this configuration setting.

#### To Do

- [ ] Create documentation for the new tag i added if that is needed in you opinion.

### CODE REVIEW

File: Resources/config/services.xml [hunk 1]
```diff
@@ -15,20 +15,38 @@
...
+        </service>
+
         <!-- Mail-->
         <service id="sulu.mail.helper" class="Sulu\Bundle\FormBundle\Mail\Helper">
             <argument type="service" id="mailer" />
             <argument>%sulu_form.mail.from%</argument>
             <argument>%sulu_form.mail.to%</argument>
             <argument>%sulu_form.mail.sender%</argument>
             <argument type="service" id="logger" />
+            <tag name="sulu_mail.handler" key="SwiftMailer" />
```
Comment (alexander-schranz) [inline Resources/config/services.xml]: I would not go here with tag I would implement it the same way we did implement the storage adapter in sulu.

File: Resources/config/services.xml [hunk 2]
```diff
@@ -15,20 +15,38 @@
...
         <service id="Sulu\Bundle\FormBundle\Form\HandlerInterface" alias="sulu_form.handler"/>
 
+        <service id="sulu.mail.helper_aware_configurator" class="Sulu\Bundle\FormBundle\Mail\HelperAwareConfigurator">
+            <argument type="service" id="sulu.mail.helper" />
+            <argument type="tagged_iterator" tag="sulu_mail.handler" index-by="key" />
+            <argument>%sulu_mail.mail.helper_name%</argument>
+        </service>
+
         <!-- Mail-->
         <service id="sulu.mail.helper" class="Sulu\Bundle\FormBundle\Mail\Helper">
```
Comment (alexander-schranz) [inline Resources/config/services.xml]: I would change here the service id to `sulu.mail.swift_mailer`

We will create an alias based on the configuration a service alias `sulu.mail.helper`for the configured service which by default will be `swift_mailer`

File: Resources/config/services.xml [hunk 3]
```diff
@@ -15,20 +15,38 @@
...
             <argument type="service" id="mailer" />
             <argument>%sulu_form.mail.from%</argument>
             <argument>%sulu_form.mail.to%</argument>
             <argument>%sulu_form.mail.sender%</argument>
             <argument type="service" id="logger" />
+            <tag name="sulu_mail.handler" key="SwiftMailer" />
+        </service>
+
+        <service id="sulu.mail.mailer_helper" class="Sulu\Bundle\FormBundle\Mail\MailerHelper">
```
Comment (alexander-schranz) [inline Resources/config/services.xml]: This should be then `sulu.mail.helper.swift_mailer.mailer`

File: Resources/config/services.xml [hunk 4]
```diff
@@ -15,20 +15,38 @@
...
+            <argument type="service" id="mailer.mailer" />
+            <argument>%sulu_form.mail.from%</argument>
+            <argument>%sulu_form.mail.to%</argument>
+            <argument>%sulu_form.mail.sender%</argument>
+            <argument type="service" id="logger" />
+            <tag name="sulu_mail.handler" key="MailerHelper" />
```
Comment (alexander-schranz) [inline Resources/config/services.xml]: Tag can be removed

File: Resources/config/services.xml [hunk 5]
```diff
@@ -15,20 +15,38 @@
...
         </service>
 
         <service id="sulu_mail.null_helper" class="Sulu\Bundle\FormBundle\Mail\NullHelper">
             <argument type="service" id="logger" />
+            <tag name="sulu_mail.handler" key="NullHelper" />
```
Comment (alexander-schranz) [inline Resources/config/services.xml]: Tag can be removed

File: Resources/config/services.xml [hunk 6]
```diff
@@ -15,20 +15,38 @@
...
+        <service id="sulu.mail.mailer_helper" class="Sulu\Bundle\FormBundle\Mail\MailerHelper">
+            <argument type="service" id="mailer.mailer" />
+            <argument>%sulu_form.mail.from%</argument>
+            <argument>%sulu_form.mail.to%</argument>
+            <argument>%sulu_form.mail.sender%</argument>
+            <argument type="service" id="logger" />
+            <tag name="sulu_mail.handler" key="MailerHelper" />
         </service>
 
         <service id="sulu_mail.null_helper" class="Sulu\Bundle\FormBundle\Mail\NullHelper">
```
Comment (alexander-schranz) [inline Resources/config/services.xml:22]: Don't see any usage for it, seems like just a service which was not used. I would `deprecate` the service definition and the whole NullLogger class itself.

File: Resources/doc/index.md [hunk 7]
```diff
@@ -46,6 +46,16 @@ sulu_form:
...
 ```
 
+Configure the email handler to SwiftMailer (default), 
+NullHelper (for no mails) or MailerHelper (for the new 
+Symfony mailer):
+
+```yml
+sulu_form:
+    mail:
+        helper: "SwiftMailer"
```
Comment (alexander-schranz) [inline Resources/doc/index.md]: Available options will then be `swift_mailer` or `mailer`

File: Tests/Application/.env [hunk 8]
```diff
@@ -2,3 +2,4 @@ APP_ENV=test
 DATABASE_URL=mysql://root:@127.0.0.1:3306/su_form_test
 DATABASE_CHARSET=utf8mb4
 DATABASE_COLLATE=utf8mb4_unicode_ci
+MAIL_HANDLER=SwiftMailer
```
Comment (alexander-schranz) [inline Tests/Application/.env]: would just go here with:

```
MAIL_HANDLER=
```

if not set it should use the default in your SuluForm Configuration class

File: Mail/MailerHelper.php [hunk 9]
```diff
@@ -0,0 +1,258 @@
...
+     * @param Email $message email message
+     * @param string $subject subject of the email
+     * @param Address[] $fromMail list of addresses already accounting for the defaults
+     * @param Address[] $toMail  list of addresses already accounting for the defaults
+     * @param Address[] $replyTo list of addresses already accounting for the defaults may be an empty array
+     * @param Address[] $ccMail list of addresses already accounting for the defaults may be an empty array
+     * @param Address[] $bccMail list of addresses already accounting for the defaults may be an empty array
+     * @param Address|null $sender address already accounting for the defaults
+     */
+    protected function setHeaders(
```
Comment (alexander-schranz) [inline Mail/MailerHelper.php]: make `private` not sure if this will change so I can not give here any BC promise

File: Mail/MailerHelper.php [hunk 10]
```diff
@@ -0,0 +1,258 @@
...
+        if ($bccMail) {
+            $message->bcc(...$bccMail);
+        }
+    }
+
+    /**
+     * @param Email $message
+     * @param \SplFileInfo[] $attachments
+     */
+    protected function setAttachments(
```
Comment (alexander-schranz) [inline Mail/MailerHelper.php]: same as above should be private

File: Mail/MailerHelper.php [hunk 11]
```diff
@@ -0,0 +1,258 @@
...
+        }
+    }
+
+    /**
+     * @param Email $message
+     * @param bool $html
+     * @param string $body
+     * @param string $plainText
+     */
+    protected function setBody(
```
Comment (alexander-schranz) [inline Mail/MailerHelper.php]: same as above should be private

File: Mail/HelperAwareInterface.php [hunk 12]
```diff
@@ -0,0 +1,10 @@
+<?php
+
+
+namespace Sulu\Bundle\FormBundle\Mail;
+
+
+interface HelperAwareInterface
```
Comment (alexander-schranz) [inline Mail/HelperAwareInterface.php]: I would not go with a AwareInterface and setter so this can be removed

File: Mail/HelperAwareConfigurator.php [hunk 13]
```diff
@@ -0,0 +1,33 @@
+<?php
+
+
+namespace Sulu\Bundle\FormBundle\Mail;
+
+
+class HelperAwareConfigurator
```
Comment (alexander-schranz) [inline Mail/HelperAwareConfigurator.php]: I would not go with a Configurator here so this can be removed.

File: Form/Handler.php [hunk 14]
```diff
@@ -83,6 +86,11 @@ public function __construct(
         $this->honeyPotField = $honeyPotField;
     }
 
+    public function setHelper(HelperInterface $mailHelper): void
```
Comment (alexander-schranz) [inline Form/Handler.php]: As written beneath we will inject the correct `Mail\HelperInterface $mailHelper` in the constructor so no setter is needed here.

File: Form/Handler.php [hunk 15]
```diff
@@ -28,7 +30,7 @@
 /**
  * Handling of form based on form configuration.
  */
-class Handler implements HandlerInterface
+class Handler implements HandlerInterface, HelperAwareInterface
```
Comment (alexander-schranz) [inline Form/Handler.php]: also this Interface can be removed

File: DependencyInjection/SuluFormExtension.php [hunk 16]
```diff
@@ -207,5 +209,7 @@ public function load(array $configs, ContainerBuilder $container)
             $container->setAlias(FormTokenController::class, 'sulu_form.form_token_controller')
                 ->setPublic(true);
         }
+
+        $container->setParameter('sulu_mail.mail.helper_name', $config['mail']['helper']);
```
Comment (alexander-schranz) [inline DependencyInjection/SuluFormExtension.php:214]: We need to check if the `mailer` component is installed so we should go here with something like:

```php
$defaultMailerHelper = null;
if (class_exists(SwiftMailer::class)) {
      $defaultMailerHelper = 'swift_mailer';
      $loader->load('swift_mailer.xml'); // this file contains the swift mailer configuration
}

if (class_exists(Mailer::class)) {
      $defaultMailerHelper = $defaultMailer ?: 'mailer';
      $loader->load('mailer.xml'); // this file contains the mailer configuration
}
```

Instead of setting a parameter here we will create a service alias:

```php
$mailHelper = $container->resolvePlaceholder($config['mail']['helper'], true);
if ('auto' === 'mailerHelper') {
      $mailerHelper = $defaultMailerHelper;
}

container->setAlias('sulu.mail.helper', 'sulu.mail.helper.' . $mailHelper)
                ->setPublic(true);

container->setAlias('sulu.mail.helper', 'sulu.mail.helper.' . $mailHelper)
                ->setPublic(true);
```

File: Mail/MailerHelper.php [hunk 17]
```diff
@@ -0,0 +1,258 @@
...
+namespace Sulu\Bundle\FormBundle\Mail;
+
+use Psr\Log\LoggerInterface;
+use Psr\Log\NullLogger;
+use Symfony\Component\HttpFoundation\File\UploadedFile;
+use Symfony\Component\Mailer\MailerInterface;
+use Symfony\Component\Mime\Address;
+use Symfony\Component\Mime\Email;
+
+class MailerHelper implements HelperInterface
```
Comment (alexander-schranz) [inline Mail/MailerHelper.php:21]: Will have a deeper look at this later.

File: Resources/config/services.xml [hunk 18]
```diff
@@ -15,20 +15,38 @@
             <argument type="service" id="sulu_media.media_manager" />
             <argument>%sulu_form.honeypot_strategy%</argument>
             <argument>%sulu_form.honeypot_field%</argument>
+            <configurator service="sulu.mail.helper_aware_configurator" />
```
Comment (alexander-schranz) [inline Resources/config/services.xml]: As written beneath this can be removed with the suggestions

File: Resources/config/services.xml [hunk 19]
```diff
@@ -15,20 +15,38 @@
...
         </service>
         <service id="Sulu\Bundle\FormBundle\Form\HandlerInterface" alias="sulu_form.handler"/>
 
+        <service id="sulu.mail.helper_aware_configurator" class="Sulu\Bundle\FormBundle\Mail\HelperAwareConfigurator">
```
Comment (alexander-schranz) [inline Resources/config/services.xml]: As written beneath this can be removed with the suggestions

File: Tests/Functional/Listener/FormRequestTest.php [hunk 20]
```diff
@@ -0,0 +1,238 @@
...
+
+use Sulu\Bundle\FormBundle\Configuration\MailConfiguration;
+use Sulu\Bundle\FormBundle\Entity\Form;
+use Sulu\Bundle\FormBundle\Entity\FormField;
+use Sulu\Bundle\FormBundle\Entity\FormFieldTranslation;
+use Sulu\Bundle\FormBundle\Entity\FormTranslation;
+use Sulu\Bundle\FormBundle\Entity\FormTranslationReceiver;
+use Sulu\Bundle\TestBundle\Testing\SuluTestCase;
+
+class FormRequestTest extends SuluTestCase
```
Comment (alexander-schranz) [inline Tests/Functional/Listener/FormRequestTest.php]: Nice 👍

File: Tests/Functional/Listener/FormRequestTest.php [hunk 21]
```diff
@@ -0,0 +1,238 @@
...
+        $mailCollector = $this->client->getProfile()->getCollector('swiftmailer');
+        // 2 messages should be send 1 to admin and 1 to email
+        $this->assertSame(0, $mailCollector->getMessageCount());
+
+        $this->assertEmailCount(0);
+    }
+
+    public function testUsesMailerWhenMailerHandlerSet()
+    {
+        $_ENV['MAIL_HANDLER'] = 'MailerHelper';
```
Comment (alexander-schranz) [inline Tests/Functional/Listener/FormRequestTest.php]: would not set a `$_ENV` here would create seperate TestCase for the `Mailer` and set there the KernelConfiguration `environment`: https://github.com/sulu/sulu/blob/3b41b627fbc1cf93be7f98eb3e3d4ebe35883b35/src/Sulu/Bundle/TestBundle/Testing/KernelTrait.php#L66-L69

to something like `test_mailer` and then in then create a `test_mailer/config.yaml` or `config_test_mailer.yaml` which is loaded by the test kernel where:

```
sulu_form:
     mail:
           helper: 'mailer'
```

is configured

File: Tests/Functional/Listener/FormRequestTest.php [hunk 22]
```diff
@@ -0,0 +1,238 @@
...
+            'url' => '/'
+        ]);
+
+        $suluDocumentManager->publish($homePage, 'de');
+        $suluDocumentManager->persist($homePage, 'de');
+        $suluDocumentManager->flush();
+    }
+
+
+    private function createFullForm(): Form
```
Comment (alexander-schranz) [inline Tests/Functional/Listener/FormRequestTest.php]: Would move this into a `CreateFormTrait` in Tests/Functional/Trait namespace.

File: Tests/Functional/Listener/FormRequestTest.php [hunk 23]
```diff
@@ -0,0 +1,238 @@
...
+            $formName . '[email]' => 'test@example.org',
+            $formName . '[email1]' => 'jon@example.org',
+        ]);
+
+        $this->client->enableProfiler();
+        $this->client->submit($formElm);
+        $this->assertResponseRedirects('?send=true');
+    }
+
+    private function inititalize(): void
```
Comment (alexander-schranz) [inline Tests/Functional/Listener/FormRequestTest.php]: with splitting this into multiple TestCase this can now be the `setUp`

File: Tests/Unit/Mail/MailerHelperTest.php [hunk 24]
```diff
@@ -0,0 +1,218 @@
...
+    private $mailerMock;
+
+    /**
+     * @var MockObject|LoggerInterface
+     */
+    private $loggerMock;
+
+    protected function setUp(): void
+    {
+        $this->mailerMock = $this->getMockBuilder(MailerInterface::class)->getMock();
```
Comment (alexander-schranz) [inline Tests/Unit/Mail/MailerHelperTest.php]: We use `prophecy` to for mocking so I would keep that. Ping me or contact me on slack if you need any help.

File: composer.json [hunk 25]
```diff
@@ -17,18 +17,20 @@
...
         "sulu/sulu": "^2.0.3 || ^2.1@dev",
         "symfony/config": "^4.3 || ^5.0",
         "symfony/console": "^4.3 || ^5.0",
         "symfony/dependency-injection": "^4.3 || ^5.0",
         "symfony/event-dispatcher": "^4.3 || ^5.0",
         "symfony/event-dispatcher-contracts": "^1.0 || ^2.0",
         "symfony/form": "^4.3 || ^5.0",
-        "symfony/validator": "^4.3 || ^5.0",
+        "symfony/mime": "^4.4 || ^5.1",
+        "symfony/mailer": "^4.4 || ^5.1",
```
Comment (alexander-schranz) [inline composer.json]: We should not increase the dependency tree and as sulu ships currently with swiftmailer. So currently this should be just `dev` dependencies and if somebody want to use the `Mailer` implementation he need to install the symfony mailer component.

**Review by alexander-schranz — CHANGES_REQUESTED** [review]
@bramtervoort Thank you for working on this 🚀 . I think we should use the same pattern which we currently use in sulu for this kind of things, else it is getting confused having different implementation for the same kind of thing. I added some comments and did not yet have enough time to have a deeper look at the tests and the implementation but wanted to give you feedback as soon as it was possible.

**Comment by bramtervoort** [comment]
Thank you Alexander for your feedback. It will be a couple of day's before I have the time to resolve them properly.

**Comment by alexander-schranz** [comment]
@bramtervoort no problem. Feel free to ping me here or over slack if you need any help.

**Comment by bramtervoort** [comment]
i made some progress with this. next weekend ill be finishing this.

File: Tests/Functional/Listener/FormRequestTest.php (hunk 22 above)
Comment (bramtervoort) [reply Tests/Functional/Listener/FormRequestTest.php]: I turned this into a fixture instead since I think that is more close to what it should be. Of course happy to still make it into a trait.

I remember there was a nice way to apply fixtures but well i came up with this :)

**Comment by bramtervoort** [comment]
There all refactored, i just love that `prophecy`. I agree it looks much better now. 

I refactored the form to a fixture if you don't agree I will be happy to turn it into a trait of course. Took some time to finish this all but i hope you can still have a look at it. Happy to resolve more review issues.

File: DependencyInjection/Configuration.php [hunk 26]
```diff
@@ -22,6 +23,10 @@
  */
 class Configuration implements ConfigurationInterface
 {
+    const NULL_HELPER = 'NullHelper';
```
Comment (alexander-schranz) [inline DependencyInjection/Configuration.php]: should be removed as null://localhost should be configured then as MAILER DSN and adding here the null helper into the new configuration doesn't make any sense at this state

File: DependencyInjection/Configuration.php [hunk 27]
```diff
@@ -22,6 +23,10 @@
...
+    const SWIFT_MAILER_HELPER = 'SwiftMailer';
```
Comment (alexander-schranz) [inline DependencyInjection/Configuration.php]:
```suggestion
    const SWIFT_MAILER_HELPER = 'swift_mailer';
```

File: DependencyInjection/Configuration.php [hunk 28]
```diff
@@ -22,6 +23,10 @@
...
+    const MAILER_HELPER = 'Mailer';
```
Comment (alexander-schranz) [inline DependencyInjection/Configuration.php]:
```suggestion
    const MAILER_HELPER = 'mailer';
```

File: DependencyInjection/SuluFormExtension.php [hunk 29]
```diff
@@ -207,5 +211,25 @@ public function load(array $configs, ContainerBuilder $container)
...
+        }
+        $container->setParameter('sulu.mail.helper', $helper);
+
+        if ($helper == Configuration::MAILER_HELPER) {
+            $container->setAlias('sulu.mail.helper', 'sulu.mail.mailer');
+        } elseif ($helper == Configuration::NULL_HELPER) {
+            $container->setAlias('sulu.mail.helper', 'sulu_mail.null_helper');
+        } elseif ($helper == Configuration::SWIFT_MAILER_HELPER) {
+            $container->setAlias('sulu.mail.helper', 'sulu.mail.swift_mailer');
+        }
```
Comment (alexander-schranz) [inline DependencyInjection/SuluFormExtension.php]: This can be simplified to:

```suggestion
        $container->setAlias('sulu.mail.helper', 'sulu.mail.' . $helper);
```

This make it also extendable for the third party provide there own mailer.

File: DependencyInjection/SuluFormExtension.php [hunk 30]
```diff
@@ -207,5 +211,25 @@ public function load(array $configs, ContainerBuilder $container)
...
+    }
+
+
+    private function configureHelper(array $config, ContainerBuilder $container)
+    {
+        $helper = $config['mail']['helper'];
+        if (\method_exists($container, 'resolveEnvPlaceholders')) {
+            $helper = $container->resolveEnvPlaceholders($helper, true);
+        }
+        $container->setParameter('sulu.mail.helper', $helper);
```
Comment (alexander-schranz) [inline DependencyInjection/SuluFormExtension.php]: This parameter should be removed. It is not used by us and the dev should not have to care about which mailer is used to send the emails.

File: Form/Builder.php [hunk 31]
```diff
@@ -272,4 +274,5 @@ private function getWebspaceKey(): ?string
 
         return $webspaceKey;
     }
+
```
Comment (alexander-schranz) [inline Form/Builder.php]: should be removed

File: Event/RequestListener.php [hunk 32]
```diff
@@ -64,7 +65,6 @@ public function onKernelRequest(RequestEvent $event): void
         }
 
         $request = $event->getRequest();
-
```
Comment (alexander-schranz) [inline Event/RequestListener.php]: would not remove this empty line :)

File: DependencyInjection/Configuration.php [hunk 33]
```diff
@@ -51,6 +56,9 @@ public function getConfigTreeBuilder()
             ->arrayNode('mail')
                 ->addDefaultsIfNotSet()
                 ->children()
+                    ->scalarNode('helper')
+                        ->defaultValue(self::SWIFT_MAILER_HELPER)
+                        ->end()
```
Comment (alexander-schranz) [inline DependencyInjection/Configuration.php]: end should be on the same level as the opening node.

```suggestion
                    ->end()
```

File: Mail/MailerHelper.php [hunk 34]
```diff
@@ -0,0 +1,258 @@
...
+            $toMail ?? $this->toMail,
+            $replyTo,
+            $subject ?? '',
+            $ccMail,
+            $bccMail,
+            $plainText
+        );
+
+        $this->mailer->send($message);
+        return 0;
```
Comment (alexander-schranz) [inline Mail/MailerHelper.php:104]: Empty line should be here before the return.

```suggestion

        return 0;
```

File: Mail/MailerHelper.php (hunk 9 above)
Comment (alexander-schranz) [reply Mail/MailerHelper.php]: still valid this should be private as we can currently not guarente a BC here.

File: Mail/MailerHelper.php (hunk 10 above)
Comment (alexander-schranz) [reply Mail/MailerHelper.php]: same as above still valid the protected function should be private

File: Mail/MailerHelper.php (hunk 11 above)
Comment (alexander-schranz) [reply Mail/MailerHelper.php]: same as above still valid the protected function should be private we can not give bc support for them here.

File: Resources/config/services.xml (hunk 19 above)
Comment (alexander-schranz) [reply Resources/config/services.xml]: Should still be valid this should be able to be removed with setting the alias like we do it in the other cases.

File: Resources/config/services.xml [hunk 35]
```diff
@@ -18,17 +18,34 @@
...
+            <argument type="service" id="sulu.mail.helper" />
+            <argument type="tagged_iterator" tag="sulu_mail.handler" index-by="key" />
+            <argument>%sulu_mail.mail.helper_name%</argument>
+        </service>
+
         <!-- Mail-->
-        <service id="sulu.mail.helper" class="Sulu\Bundle\FormBundle\Mail\Helper">
+        <service id="sulu.mail.helper" alias="sulu.mail.swift_mailer" />
```
Comment (alexander-schranz) [inline Resources/config/services.xml]: This is set in the configuration so this don't need to be set here again. So we can remove this.

File: Resources/config/services.xml [hunk 36]
```diff
@@ -18,17 +18,34 @@
...
+            <argument type="service" id="mailer.mailer" />
+            <argument>%sulu_form.mail.from%</argument>
+            <argument>%sulu_form.mail.to%</argument>
+            <argument>%sulu_form.mail.sender%</argument>
+            <argument type="service" id="logger" />
+        </service>
+
         <service id="sulu_mail.null_helper" class="Sulu\Bundle\FormBundle\Mail\NullHelper">
             <argument type="service" id="logger" />
+            <tag name="sulu_mail.handler" key="NullHelper" />
```
Comment (alexander-schranz) [inline Resources/config/services.xml]: Tag can be removed but service should also be marked as deprecated as it was actually never used:

```suggestion
           <deprecated />
```

File: Resources/doc/index.md [hunk 37]
```diff
@@ -46,6 +46,16 @@ sulu_form:
         sender: "%env(SULU_ADMIN_EMAIL)%"
 ```
 
+Configure the email handler to SwiftMailer (default),
```
Comment (alexander-schranz) [inline Resources/doc/index.md]: Configure the email handler to SwiftMailer (`swift_mailer`) or the Symfony Mailer (`mailer`)

File: Resources/doc/index.md (hunk 7 above)
Comment (alexander-schranz) [reply Resources/doc/index.md]:
```suggestion
        helper: "swift_mailer" # is default
```

File: SuluFormBundle.php [hunk 38]
```diff
@@ -13,8 +13,10 @@
 
 use Sulu\Bundle\FormBundle\DependencyInjection\CompilerPass\DynamicListBuilderCompilerPass;
 use Sulu\Bundle\FormBundle\DependencyInjection\CompilerPass\ListProviderCompilerPass;
+use Sulu\Bundle\FormBundle\DependencyInjection\CompilerPass\MailHelperCompilerPass;
```
Comment (alexander-schranz) [inline SuluFormBundle.php]: looks like unused use statements here

File: Tests/Application/Kernel.php [hunk 39]
```diff
@@ -11,9 +11,13 @@
 
 namespace Sulu\Bundle\FormBundle\Tests\Application;
 
+use Sulu\Bundle\FormBundle\Tests\Application\Controller\DefaultController;
```
Comment (alexander-schranz) [inline Tests/Application/Kernel.php]: looks like unused use statements here

File: Tests/Application/MailerKernel.php [hunk 40]
```diff
@@ -0,0 +1,26 @@
+<?php
+
+
+namespace Sulu\Bundle\FormBundle\Tests\Application;
```
Comment (alexander-schranz) [inline Tests/Application/MailerKernel.php:12]: code style should be fixed and the copyright header should be added

File: Tests/Application/MailerKernel.php [hunk 41]
```diff
@@ -0,0 +1,26 @@
...
+        $loader->load(__DIR__ . '/config/config_test_mailer.yaml');
+    }
+
+    public function getCacheDir()
+    {
+        return $this->getProjectDir() . \DIRECTORY_SEPARATOR
+            . 'var' . \DIRECTORY_SEPARATOR
+            . 'cache' . \DIRECTORY_SEPARATOR
+            . $this->getContext()  . \DIRECTORY_SEPARATOR
+            . $this->environment;
```
Comment (alexander-schranz) [inline Tests/Application/MailerKernel.php]: this should have its own cache directory e.g. `$this->environment . '_mailer'` else it could conflict with the other Kernel.

File: Tests/Application/config/routing_website.yml [hunk 42]
```diff
@@ -1,2 +0,0 @@
-sulu_form:
```
Comment (alexander-schranz) [inline Tests/Application/config/routing_website.yml:1]: why was this removed?

File: Tests/Application/config/config_test_mailer.yaml [hunk 43]
```diff
@@ -0,0 +1,6 @@
+#import:
+#    - config.yml
+#
+sulu_form:
+    mail:
+        helper: 'Mailer'
```
Comment (alexander-schranz) [inline Tests/Application/config/config_test_mailer.yaml]:
```suggestion
sulu_form:
    mail:
        helper: 'mailer'
```

File: Tests/Functional/Mail/HelperTestCase.php [hunk 44]
```diff
@@ -0,0 +1,107 @@
...
+        $this->client = $this->createWebsiteClient();
+        $this->purgeDatabase();
+        $this->initPhpcr();
+        $this->entityManager = $this->getEntityManager();
+
+        $fixture = new LoadFormFixture();
+        $fixture->load($this->entityManager);
+    }
+
+//    public function testPageRendersWithoutForm()
```
Comment (alexander-schranz) [inline Tests/Functional/Mail/HelperTestCase.php]: should be removed when not used

File: Tests/Functional/Mail/Fixtures/LoadFormFixture.php [hunk 45]
```diff
@@ -0,0 +1,117 @@
...
+        $formTranslation->addReceiver($ccReceiver);
+        $formTranslation->addReceiver($bccReceiver);
+
+        $formTranslation->setForm($form);
+        $form->addTranslation($formTranslation);
+
+        $formField->setForm($form);
+
+        $manager->persist($form);
+        $manager->flush();
```
Comment (alexander-schranz) [inline Tests/Functional/Mail/Fixtures/LoadFormFixture.php]: We should not flush here. the flush and clear should do the test case itself as maybe there other entities need to be flushed with it and as flush has a big impact on performance we should not use only use it when it is really needed.

File: Tests/Functional/Mail/Fixtures/LoadFormFixture.php [hunk 46]
```diff
@@ -0,0 +1,117 @@
...
+        $manager->clear();
+    }
+
```
Comment (alexander-schranz) [inline Tests/Functional/Mail/Fixtures/LoadFormFixture.php]: code style should also be fixed here.

File: Tests/Functional/Mail/HelperTest.php [hunk 47]
```diff
@@ -0,0 +1,32 @@
+<?php
+
```
Comment (alexander-schranz) [inline Tests/Functional/Mail/HelperTest.php:2]: PHPCS should be run here not sure why php cs was not run for this PR. this files should be fixed when run:

```
composer php-cs-fix
```

File: Tests/Unit/Mail/MailerHelperTest.php [hunk 48]
```diff
@@ -0,0 +1,210 @@
...
+use Symfony\Component\HttpFoundation\File\UploadedFile;
+use Symfony\Component\Mailer\MailerInterface;
+use Symfony\Component\Mime\Email;
+use Symfony\Component\Mime\Address;
+
+/**
+ * Class MailerHelperTest
+ * This tests the MailerHelper class that sends an email using the symfony mailer component.
+ *
+ * The MailerHelper class has 3 protected methods that could be overwritten to change the mail message.
```
Comment (alexander-schranz) [inline Tests/Unit/Mail/MailerHelperTest.php]: I will currently not support this. If somebody want to hook in he should use the official symfony way, listening on events of symfony mailer or create a swift mailer plugin.

File: composer.json (hunk 25 above)
Comment (alexander-schranz) [inline composer.json]: This should still be "dev" dependencies at current state mailer should not be shipped with sulu form bundle. When somebody want to switch to the mailer it should actively require it in there composer.json file.

Also it should be here `^4.4 || ^5.0`

**Review by alexander-schranz — COMMENTED** [review]
@bramtervoort Thx for working on this. I added some more comments about the actual implementation.

**Comment by alexander-schranz** [comment]
@bramtervoort can you also rebase this branch to solve the conflicts and can you try to update your forks branches. I'm not sure why github action are not started for this PR but maybe it has todo that the fork branches are not up2date think it is worse a try.

**Comment by bramtervoort** [comment]
> @bramtervoort Thx for working on this. I added some more comments about the actual implementation.

Thx, I will check if can find a little time to work some more on this.

**Comment by bramtervoort** [comment]
There found the time to update this :D thank you for your patience. If i can do anything to fix or improve the code please let me know. ( sorry this took me so long ;) )

**Review by alexander-schranz — APPROVED** [review]

**Comment by alexander-schranz** [comment]
@bramtervoort Thank you for contributing this 👍
